base_layer/core/src/base_node/states/horizon_state_sync/validators/chain_balance.rs

```diff
@@ -0,0 +1,129 @@
+//  Copyright 2020, The Tari Project
+//
+//  Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+//  following conditions are met:
+//
+//  1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+//  disclaimer.
+//
+//  2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+//  following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+//  3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+//  products derived from this software without specific prior written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+//  INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+//  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+//  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use crate::{
+    base_node::states::horizon_state_sync::validators::header_iter::HeaderIter,
+    chain_storage::{BlockchainBackend, BlockchainDatabase},
+    consensus::ConsensusManager,
+    transactions::{
+        tari_amount::MicroTari,
+        types::{BlindingFactor, Commitment, CryptoFactories, PrivateKey},
+    },
+    validation::{StatelessValidation, ValidationError},
+};
+use log::*;
+use tari_crypto::commitment::HomomorphicCommitmentFactory;
+
+const LOG_TARGET: &str = "c::bn::states::horizon_state_sync::chain_balance";
+
+/// Validate that the chain balances at a given height.
+pub struct ChainBalanceValidator<B> {
+    rules: ConsensusManager,
+    db: BlockchainDatabase<B>,
+    factories: CryptoFactories,
+}
+
+impl<B: BlockchainBackend> ChainBalanceValidator<B> {
+    pub fn new(db: BlockchainDatabase<B>, rules: ConsensusManager, factories: CryptoFactories) -> Self {
+        Self { db, rules, factories }
+    }
+}
+
+impl<B: BlockchainBackend> StatelessValidation<u64> for ChainBalanceValidator<B> {
```

Comment:
Should this not rather be statefull? And the db be passed in?

Reply:
The "stateful" validator should be removed IMO, there is no need for it as you can have a database reference in the struct if needed. That way you don't need to differentiate the type of validation and it will make the code simpler.

Reply:
I used the "Stateful" validator but somewhere along the way this made much more sense (`StatelessValidation` should just be called `Validation` - or `Validator`)